Add Apple WeatherKit integration

tjhorner opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change

This PR adds an integration for Apple's [WeatherKit API](https://developer.apple.com/documentation/weatherkitrestapi). It's the most direct replacement for the Dark Sky API (RIP) since Apple bought it and integrated Dark Sky's tech into Apple Weather.

I am Extremely Not familiar with the Python ecosystem, and especially unfamiliar with the best practices of the Home Assistant project. I did try to follow what was documented on the developers site as closely as possible, but please feel free to flag areas that need improvement.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/28833

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

tjhorner committed:
Add Apple WeatherKit integration

tjhorner committed:
Remove unnecessary comments/consts

tjhorner committed:
Add additional data set support check in coordinator

joostlek commented on homeassistant/components/weatherkit/__init__.py:
```diff
@@ -0,0 +1,49 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up this integration using UI."""
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = coordinator = WeatherKitDataUpdateCoordinator(
```
Please first create the coordinator, then execute the first refresh, only then you may add it to the hass object.

joostlek commented on homeassistant/components/weatherkit/__init__.py:
```diff
@@ -0,0 +1,49 @@
+    return unloaded
+
+
+async def async_reload_entry(hass: HomeAssistant, entry: ConfigEntry) -> None:
+    """Reload config entry."""
+    await async_unload_entry(hass, entry)
+    await async_setup_entry(hass, entry)
```
Only having `async_unload_entry` as function should be enough to enable reloading the integration so you can remove this one

bdraco replied:
```suggestion
```

joostlek commented on homeassistant/components/weatherkit/config_flow.py:
```diff
@@ -0,0 +1,122 @@
+
+    async def async_step_user(
+        self,
+        user_input: dict | None = None,
```
```suggestion
        user_input: dict[str, Any] | None = None,
```

joostlek commented on homeassistant/components/weatherkit/config_flow.py:
```diff
@@ -0,0 +1,122 @@
+                _errors["base"] = "unsupported_location"
+            except WeatherKitApiClientAuthenticationError as exception:
+                LOGGER.warning(exception)
+                _errors["base"] = "auth"
```
```suggestion
                _errors["base"] = "invalid_auth"
```
auth... what?

joostlek commented on homeassistant/components/weatherkit/config_flow.py:
```diff
@@ -0,0 +1,122 @@
+                _errors["base"] = "auth"
+            except WeatherKitApiClientCommunicationError as exception:
+                LOGGER.error(exception)
+                _errors["base"] = "connection"
```
```suggestion
                _errors["base"] = "cannot_connect"
```

joostlek commented on homeassistant/components/weatherkit/config_flow.py:
```diff
@@ -0,0 +1,122 @@
+
+DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_NAME, default="Home"): str,
```
We want to avoid having people set names themselves in new integrations/config flows and are slowly removing it in current integrations as well. People can set the name by renaming the device and or config entry.

joostlek commented on homeassistant/components/weatherkit/config_flow.py:
```diff
@@ -0,0 +1,122 @@
+        user_input: dict | None = None,
+    ) -> data_entry_flow.FlowResult:
+        """Handle a flow initialized by the user."""
+        _errors = {}
```
```suggestion
        errors = {}
```
Doesn't really need to be private imo

joostlek commented on homeassistant/components/weatherkit/config_flow.py:
```diff
@@ -0,0 +1,122 @@
+                _errors["base"] = "unknown"
+            else:
+                return self.async_create_entry(
+                    title=user_input[CONF_NAME],
```
Instead of name you could use the coordinates as name

joostlek commented on homeassistant/components/weatherkit/config_flow.py:
```diff
@@ -0,0 +1,122 @@
+            else:
+                return self.async_create_entry(
+                    title=user_input[CONF_NAME],
+                    data=user_input,
```
We like our config entry data structures flat, so it should flat out the CONF_LOCATION so the latitude and longitude are on the same level as the rest

tjhorner replied:
Got it. I don't think this was mentioned in the developer docs. Is there any particular reason for this? Easier to serialize?

joostlek replied:
Well, it would certainly work, I also got this as feedback as one of my own PRs, but I am not sure about the reasoning.

I think the main goal would be to avoid integration putting in complete data structures.

In this case, accessing one level down costs more than staying on the same level. On this scale it doesn't matter that much, but when you get complete data structures in a lot of different integrations... it would get out of hand.

Note: we do use an object in oauth integrations to store token related data. Why this is chosen, not sure, but just to let you know that the story above isn't 100% applicable.

joostlek commented on homeassistant/components/weatherkit/config_flow.py:
```diff
@@ -0,0 +1,122 @@
+            errors=_errors,
+        )
+
+    async def _test_config(self, user_input) -> None:
```
```suggestion
    async def _test_config(self, user_input: dict[str, Any]) -> None:
```

joostlek commented on homeassistant/components/weatherkit/coordinator.py:
```diff
@@ -0,0 +1,75 @@
+
+    config_entry: ConfigEntry
+    supported_data_sets: list[
+        Literal["currentWeather", "forecastDaily", "forecastHourly"]
```
Can these be made constants or moved to the library? (not sure whats best here)

tjhorner replied:
This is something I was torn about too. I think since that these values are closely related to the API details (they are the actual keys that get sent to WeatherKit), I can abstract it by creating an enum in the library for the possible data sets, and use that when passing data to and from the library. That way the client isn't concerned with the underlying values, and we have nice types.

joostlek replied:
Sounds beautiful

joostlek commented on homeassistant/components/weatherkit/coordinator.py:
```diff
@@ -0,0 +1,75 @@
+    ) -> None:
+        """Initialize."""
+        self.client = client
+        self.supported_data_sets = None
```
This could be already initialized above

joostlek commented on homeassistant/components/weatherkit/coordinator.py:
```diff
@@ -0,0 +1,75 @@
+                self.supported_data_sets,
+            )
+        except WeatherKitApiClientAuthenticationError as exception:
+            raise ConfigEntryAuthFailed(exception) from exception
```
This would trigger a reauth flow, but you don't have any. Can we check if we can actually get data in the `__init__.py`'s `async_setup_entry` and raise the ConfigEntryAuthFailed there?

tjhorner replied:
What is the difference throwing it there? Not triggering the re-auth flow?

joostlek replied:
Good question, not sure atm

bdraco replied:
Either method is going to trigger the reauth flow which will fail since it doesn't exist.

You can manually test the credentials in `__init__.py:async_setup_entry`, log an error, and return `False` if it fails for now.

In a future PR, change it to `raise ConfigEntryAuthFailed`

Also there is no need to call `ConfigEntryAuthFailed(exception)` when you do `raise ConfigEntryAuthFailed from exception` since it will figure out how to handle it from the cause

joostlek commented on homeassistant/components/weatherkit/manifest.json:
```diff
@@ -0,0 +1,13 @@
+  "name": "Apple WeatherKit",
+  "codeowners": ["@tjhorner"],
+  "config_flow": true,
+  "dependencies": [],
+  "documentation": "https://www.home-assistant.io/integrations/weatherkit",
+  "homekit": {},
+  "iot_class": "cloud_polling",
+  "requirements": ["apple_weatherkit==0.2.0"],
+  "ssdp": [],
+  "zeroconf": []
```
```suggestion
  "documentation": "https://www.home-assistant.io/integrations/weatherkit",
  "iot_class": "cloud_polling",
  "requirements": ["apple_weatherkit==0.2.0"]
```

joostlek commented on homeassistant/components/weatherkit/strings.json:
```diff
@@ -0,0 +1,25 @@
+  "config": {
+    "step": {
+      "user": {
+        "title": "WeatherKit Setup",
```
```suggestion
        "title": "WeatherKit setup",
```

joostlek commented on homeassistant/components/weatherkit/strings.json:
```diff
@@ -0,0 +1,25 @@
+        "description": "Enter your location details and WeatherKit authentication credentials below.",
+        "data": {
+          "name": "Name",
+          "location": "Location",
```
```suggestion
          "location": "[%key:common::config_flow::data::location%]",
```

tjhorner replied:
Is there somewhere I could find all the common translation keys?

joostlek commented on homeassistant/components/weatherkit/strings.json:
```diff
@@ -0,0 +1,25 @@
+          "name": "Name",
+          "location": "Location",
+          "key_id": "Key ID",
+          "team_id": "Apple Team ID",
```
```suggestion
          "team_id": "Apple team ID",
```
Only if "Apple Team" is a brand, otherwise ignore

joostlek commented on homeassistant/components/weatherkit/strings.json:
```diff
@@ -0,0 +1,25 @@
+          "key_id": "Key ID",
+          "team_id": "Apple Team ID",
+          "service_id": "Service ID",
+          "key_pem": "Private Key (.p8)"
```
```suggestion
          "key_pem": "Private key (.p8)"
```

joostlek commented on homeassistant/components/weatherkit/strings.json:
```diff
@@ -0,0 +1,25 @@
+      }
+    },
+    "error": {
+      "already_configured": "This integration has already been configured.",
```
```suggestion
      "already_configured": "[%key:common::config_flow::abort::already_configured_location%]",
```

joostlek commented on homeassistant/components/weatherkit/weather.py:
```diff
@@ -0,0 +1,255 @@
+    def unique_id(self) -> str:
+        """Return unique ID."""
+        location = self._config[CONF_LOCATION]
+        return f"{location[CONF_LATITUDE]}-{location[CONF_LONGITUDE]}"
```
This can be a shorthand attribute in the constructor `_attr_unique_id = ...`

tjhorner replied:
I was also wondering if this unique ID should be prefixed with the domain, or if it is unique within the domain. Because I imagine users may have several entities linked to their home location.

bdraco replied:
no domain needed, its already considered

https://github.com/home-assistant/core/blob/d59aa958b6d6eda6d8aef3dc406edce1a7301459/homeassistant/components/homekit/aidmanager.py#L38

joostlek commented on homeassistant/components/weatherkit/weather.py:
```diff
@@ -0,0 +1,255 @@
+    }
+
+
+class WeatherKitWeather(
```
New integrations must use `has_entity_name = True`. My advice would be to add `_attr_has_entity_name = True`, add a device (service) (without a name) and then set `_attr_name = None`. Check IPMA for an example. This way the name of the config entry is used as name for the entity, so if someone changes the config entry name, they automatically get their weather renamed

bdraco replied:
https://developers.home-assistant.io/docs/core/entity?_highlight=has_en#has_entity_name-true-mandatory-for-new-integrations

tjhorner replied:
So, I set `_attr_name` to `None` then attached a device with `_attr_device_info`, but the sensor only gets renamed when the service (device) gets renamed, not the config entry. This is totally fine IMO, but I was wondering if I was missing something to make the config entry name match the service (and thus, the sensor's) name

joostlek replied:
Euh, afaik that should be the case I will do some tests. For now, your code is good :)

joostlek commented on homeassistant/components/weatherkit/weather.py:
```diff
@@ -0,0 +1,255 @@
+    @property
+    def native_temperature(self) -> float | None:
+        """Return the current temperature."""
+        temperature = self.coordinator.data.get("currentWeather").get("temperature")
+        return temperature
```
```suggestion
        return self.coordinator.data.get("currentWeather").get("temperature")
```

joostlek commented on homeassistant/components/weatherkit/weather.py:
```diff
@@ -0,0 +1,255 @@
+    @property
+    def native_apparent_temperature(self) -> float | None:
+        """Return the current apparent_temperature."""
+        apparent_temperature = self.coordinator.data.get("currentWeather").get(
+            "temperatureApparent"
+        )
+        return apparent_temperature
```
Idem

joostlek commented on homeassistant/components/weatherkit/weather.py:
```diff
@@ -0,0 +1,255 @@
+    def humidity(self) -> float | None:
+        """Return the current humidity."""
+        humidity = self.coordinator.data.get("currentWeather").get("humidity")
+        return humidity * 100
+
+    @property
+    def cloud_coverage(self) -> int | None:
+        """Return the current cloud_coverage."""
+        cloud_coverage = self.coordinator.data.get("currentWeather").get("cloudCover")
+        return cloud_coverage * 100
+
+    @property
+    def uv_index(self) -> float | None:
+        """Return the current uv_index."""
+        uv_index = self.coordinator.data.get("currentWeather").get("uvIndex")
+        return uv_index
+
+    @property
+    def native_visibility(self) -> float | None:
+        """Return the current visibility."""
+        visibility = self.coordinator.data.get("currentWeather").get("visibility")
+        return visibility / 1000
+
+    @property
+    def native_wind_gust_speed(self) -> float | None:
+        """Return the current wind_gust_speed."""
+        wind_gust_speed = self.coordinator.data.get("currentWeather").get("windGust")
+        return wind_gust_speed
+
+    @property
+    def native_wind_speed(self) -> float | None:
+        """Return the current wind_speed."""
+        wind_speed = self.coordinator.data.get("currentWeather").get("windSpeed")
+        return wind_speed
+
+    @property
+    def wind_bearing(self) -> float | None:
+        """Return the current wind_bearing."""
+        wind_bearing = self.coordinator.data.get("currentWeather").get("windDirection")
+        return wind_bearing
```
Idem

joostlek commented on .coveragerc:
```diff
@@ -1471,6 +1471,9 @@ omit =
     homeassistant/components/watson_tts/tts.py
     homeassistant/components/watttime/__init__.py
     homeassistant/components/watttime/sensor.py
+    homeassistant/components/weatherkit/__init__.py
+    homeassistant/components/weatherkit/weather.py
+    homeassistant/components/weatherkit/coordinator.py
```
Consider adding tests as they reduce the amount of regressions

tjhorner replied:
What's the best way to test with an integration that relies on an external service? Mocking the library?

joostlek replied:
Yes! You mock a library response and you test what the sensors and entities look like

joostlek reviewed (changes requested):
Initial feedback.

Thanks!

By the way, since you followed the dev docs closely, please let us know if you think information is missing (that arent in the docs but are now requested on review) as it helps us to get another perspective :)

bdraco commented on homeassistant/components/weatherkit/weather.py:
```diff
@@ -0,0 +1,255 @@
+        [
+            WeatherKitWeather(
+                coordinator,
+                config_entry.data,
```
It would be cleaner to put the data in the coordinator so you only have to pass around one object

bdraco commented on homeassistant/components/weatherkit/weather.py:
```diff
@@ -0,0 +1,255 @@
+    @property
+    def native_dew_point(self) -> float | None:
+        """Return the current dew_point."""
+        dew_point = self.coordinator.data.get("currentWeather").get(
```
You could make a 

```
@property
def data(self) -> `Type here`:
    """Return coordinator data."""
    return self.coordinator.data

Than you can replace all the `self.coordinator.data` calls with `self.data`

Its a tiny bit slower but makes all the properties a lot more readable

bdraco replied:
You might also want something that like this to avoid writing out all the currentWeather gets

```python
@property
def current_weather(self) -> dict[str, Any]:
    """Return current weather data."""
    return self.coordinator.data.get("currentWeather")
```

tjhorner committed:
Apply suggestions from code review (weatherkit)

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

bdraco commented on homeassistant/generated/integrations.json:
```diff
@@ -6308,6 +6308,12 @@
       "config_flow": true,
       "iot_class": "cloud_polling"
     },
+    "weatherkit": {
+      "name": "Apple WeatherKit",
+      "integration_type": "hub",
+      "config_flow": true,
+      "iot_class": "cloud_polling"
```
Should be added to the apple brand as well

tjhorner replied:
Where would I add that? Since this is a generated file I assume I'd need to add it elsewhere, but I am looking in the manifests of other integrations in the Apple brand (`homekit`, `homekit_controller`) and I'm not seeing anything there

tjhorner replied:
Nevermind, found the relevant documentation: https://developers.home-assistant.io/docs/creating_integration_brand

tjhorner committed:
Additional changes from code review (weatherkit)

Add service (device) to entity; remove name from user input config; flatten location in config entry

tjhorner committed:
Add weatherkit to Apple brand

tjhorner committed:
Obtain config from coordinator config_entry (weatherkit)

tjhorner committed:
Switch to using enum for data set type (weatherkit)

tjhorner committed:
Add some tests for weather entity (weatherkit)

bdraco commented on tests/components/weatherkit/test_weather.py:
```diff
@@ -0,0 +1,83 @@
+    state = hass.states.get("weather.home")
+    assert state
+    assert state.state == "partlycloudy"
+    assert state.attributes.get(ATTR_WEATHER_HUMIDITY) == 91
+    assert state.attributes.get(ATTR_WEATHER_PRESSURE) == 1009.8
+    assert state.attributes.get(ATTR_WEATHER_TEMPERATURE) == 22.9
+    assert state.attributes.get(ATTR_WEATHER_VISIBILITY) == 20.97
+    assert state.attributes.get(ATTR_WEATHER_WIND_BEARING) == 259
+    assert state.attributes.get(ATTR_WEATHER_WIND_SPEED) == 5.23
+    assert state.attributes.get(ATTR_WEATHER_APPARENT_TEMPERATURE) == 24.9
+    assert state.attributes.get(ATTR_WEATHER_DEW_POINT) == 21.3
+    assert state.attributes.get(ATTR_WEATHER_CLOUD_COVERAGE) == 62
+    assert state.attributes.get(ATTR_WEATHER_WIND_GUST_SPEED) == 10.53
+    assert state.attributes.get(ATTR_WEATHER_UV_INDEX) == 1
+    assert state.attributes.get(ATTR_ATTRIBUTION) == ATTRIBUTION
```
Please use `[]` instead of `.get()` when the attribute is expected to be there

tjhorner committed:
Improve tests for weatherkit, change supported_features detection

tjhorner committed:
Change where authentication is checked and error differently (weatherkit)

bdraco commented on homeassistant/components/weatherkit/__init__.py:
```diff
@@ -28,6 +39,15 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
 
+    try:
+        await coordinator.update_supported_data_sets()
+    except (
+        WeatherKitApiClientAuthenticationError,
+        WeatherKitApiClientError,
+    ):
+        LOGGER.exception("Error initializing integration")
+        return False
```
```suggestion
    except WeatherKitApiClientAuthenticationError:
        LOGGER.error("Authentication error initializing integration")
        return False
    except WeatherKitApiClientError as ex:
        raise ConfigEntryNotReady from ex
```

joostlek replied:
Why do we return False instead of raise ConfigEntryNotReady?

bdraco replied:
Authentication failures are not recoverable without reauth being implemented, but transient API errors should we try

tjhorner replied:
I was under the impression that `LOGGER.exception` logged exception information automatically. I looked at other uses and they didn't pass in the exception manually.

bdraco replied:
It does but we don't want the noise of a whole traceback in the log as we already know why it failed.

tjhorner replied:
I see, thanks!

tjhorner committed:
Apply suggestion from code review (weatherkit)

Co-authored-by: J. Nick Koston <nick@koston.org>

tjhorner committed:
Add missing import

bdraco commented on homeassistant/components/weatherkit/config_flow.py:
```diff
@@ -0,0 +1,126 @@
+                errors["base"] = "unknown"
+            else:
+                # Flatten location
+                user_input[CONF_LATITUDE] = user_input[CONF_LOCATION][CONF_LATITUDE]
+                user_input[CONF_LONGITUDE] = user_input[CONF_LOCATION][CONF_LONGITUDE]
+                del user_input[CONF_LOCATION]
```
```suggestion
                location = user_input.pop(CONF_LOCATION)
                user_input[CONF_LATITUDE] = location[CONF_LATITUDE]
                user_input[CONF_LONGITUDE] = location[CONF_LONGITUDE]
```

untested but this looks a bit cleaner

tjhorner replied:
Agreed, will test and see if it works

bdraco commented on homeassistant/components/weatherkit/weather.py:
```diff
@@ -0,0 +1,255 @@
+        config_entry.entry_id
+    ]
+
+    async_add_entities(
+        [
+            WeatherKitWeather(
+                coordinator,
+            ),
+        ]
+    )
```
```suggestion
    async_add_entities([WeatherKitWeather(coordinator)])
```

bdraco commented:
This looks close to done. started the CI

joostlek commented on homeassistant/components/weatherkit/coordinator.py:
```diff
@@ -0,0 +1,70 @@
+"""DataUpdateCoordinator for WeatherKit integration. Updated every 15 minutes."""
```
```suggestion
"""DataUpdateCoordinator for WeatherKit integration."""
```

joostlek commented on homeassistant/components/weatherkit/weather.py:
```diff
@@ -0,0 +1,255 @@
+    ) -> None:
+        """Initialise the platform with a data instance and site."""
+        super().__init__(coordinator)
+        self._config = coordinator.config_entry.data
+        self._attr_unique_id = (
+            f"{self._config[CONF_LATITUDE]}-{self._config[CONF_LONGITUDE]}"
+        )
```
```suggestion
        config_data = coordinator.config_entry.data
        self._attr_unique_id = (
            f"{config_data[CONF_LATITUDE]}-{config_data[CONF_LONGITUDE]}"
        )
```
Looks like you're never using `self._config` outside of the constructor.

tjhorner replied:
Ah, yeah. I think I used it elsewhere in a previous revision but forgot to remove it

joostlek commented on tests/components/weatherkit/__init__.py:
```diff
@@ -0,0 +1,70 @@
+            CONF_TEAM_ID: "ABCD123456",
+            CONF_KEY_PEM: "-----BEGIN PRIVATE KEY-----\nwhateverkey\n-----END PRIVATE KEY-----",
+        },
+        pref_disable_polling=True,
```
Why do we disable polling?

tjhorner replied:
I think I was worried about making excessive API calls, but that doesn't make sense because we're mocking the responses. So I guess it's fine to remove this line.

joostlek commented on tests/components/weatherkit/test_config_flow.py:
```diff
@@ -0,0 +1,137 @@
+
+async def _test_exception_generates_error(
+    hass: HomeAssistant, exception: Exception, error: str
+):
```
```suggestion
) -> None:
```

joostlek commented on tests/components/weatherkit/test_config_flow.py:
```diff
@@ -0,0 +1,137 @@
+        "apple_weatherkit.client.WeatherKitApiClient.get_availability",
+        side_effect=exception,
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            example_user_input,
+        )
+
+    assert result2["type"] == FlowResultType.FORM
+    assert result2["errors"] == {"base": error}
```
```suggestion
        result = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            example_user_input,
        )

    assert result["type"] == FlowResultType.FORM
    assert result["errors"] == {"base": error}
```

joostlek commented on tests/components/weatherkit/test_config_flow.py:
```diff
@@ -0,0 +1,137 @@
+        "homeassistant.components.weatherkit.config_flow.WeatherKitFlowHandler._test_config",
+        return_value=None,
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            example_user_input,
+        )
+        await hass.async_block_till_done()
+
+    assert result2["type"] == FlowResultType.CREATE_ENTRY
+
+    location = example_user_input[CONF_LOCATION]
+    assert result2["title"] == f"{location[CONF_LATITUDE]}, {location[CONF_LONGITUDE]}"
+
+    assert result2["data"] == example_config_data
```
```suggestion
        result = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            example_user_input,
        )
        await hass.async_block_till_done()

    assert result["type"] == FlowResultType.CREATE_ENTRY

    location = example_user_input[CONF_LOCATION]
    assert result["title"] == f"{location[CONF_LATITUDE]}, {location[CONF_LONGITUDE]}"

    assert result["data"] == example_config_data
```

joostlek commented on tests/components/weatherkit/test_config_flow.py:
```diff
@@ -0,0 +1,137 @@
+    assert len(mock_setup_entry.mock_calls) == 1
+
+
+async def test_form_invalid_auth(hass: HomeAssistant) -> None:
+    """Test we handle invalid auth."""
+    await _test_exception_generates_error(
+        hass, WeatherKitApiClientAuthenticationError, "invalid_auth"
+    )
+
+
+async def test_form_cannot_connect(hass: HomeAssistant) -> None:
+    """Test we handle connection errors."""
+    await _test_exception_generates_error(
+        hass, WeatherKitApiClientCommunicationError, "cannot_connect"
+    )
```
You can use `@pytest.mark.parametrize` to create these tests quickly. Checkout the opensky test `test_options_flow_failures` for an example.

joostlek commented on tests/components/weatherkit/test_config_flow.py:
```diff
@@ -0,0 +1,137 @@
+        "apple_weatherkit.client.WeatherKitApiClient.get_availability",
+        return_value=[],
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            example_user_input,
+        )
+
+    assert result2["type"] == FlowResultType.FORM
+    assert result2["errors"] == {"base": "unsupported_location"}
```
```suggestion
        result = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            example_user_input,
        )

    assert result["type"] == FlowResultType.FORM
    assert result["errors"] == {"base": "unsupported_location"}
```

joostlek commented on tests/components/weatherkit/test_config_flow.py:
```diff
@@ -0,0 +1,137 @@
+            example_user_input,
+        )
+
+    assert result2["type"] == FlowResultType.FORM
```
Please make sure all tests are ending with either a succesfull added config entry or an abort. This way we test if a config flow can recover from a failure. (So to test it will succeed when you enter a valid location after you entered an unsupported location)

tjhorner committed:
Improvements based on code review

tjhorner committed:
Improve tests and code coverage

tjhorner committed:
Additional changes from code review

bdraco commented on homeassistant/components/weatherkit/coordinator.py:
```diff
@@ -0,0 +1,70 @@
+            update_interval=timedelta(minutes=15),
+        )
+
+    async def update_supported_data_sets(self):
+        """Obtain the supported data sets for this location and store them."""
+        supported_data_sets = await self.client.get_availability(
+            self.config_entry.data[CONF_LATITUDE],
+            self.config_entry.data[CONF_LONGITUDE],
+        )
+
+        requested_data_sets = [
+            DataSetType.CURRENT_WEATHER,
+            DataSetType.DAILY_FORECAST,
+            DataSetType.HOURLY_FORECAST,
+        ]
+
+        self.supported_data_sets = [
+            data_set
+            for data_set in requested_data_sets
+            if data_set in supported_data_sets
+        ]
+
+        LOGGER.debug("Supported data sets: %s", self.supported_data_sets)
```
Lets make this a constant so it doesn't construct a new list every time this runs
```diff
diff --git a/homeassistant/components/weatherkit/coordinator.py b/homeassistant/components/weatherkit/coordinator.py
index 3675ed32eb2..29402fa7617 100644
--- a/homeassistant/components/weatherkit/coordinator.py
+++ b/homeassistant/components/weatherkit/coordinator.py
@@ -13,6 +13,12 @@ from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, Upda
 
 from .const import DOMAIN, LOGGER
 
+REQUESTED_DATA_SETS = (
+    DataSetType.CURRENT_WEATHER,
+    DataSetType.DAILY_FORECAST,
+    DataSetType.HOURLY_FORECAST,
+)
+
 
 class WeatherKitDataUpdateCoordinator(DataUpdateCoordinator):
     """Class to manage fetching data from the API."""
@@ -40,19 +46,11 @@ class WeatherKitDataUpdateCoordinator(DataUpdateCoordinator):
             self.config_entry.data[CONF_LATITUDE],
             self.config_entry.data[CONF_LONGITUDE],
         )
-
-        requested_data_sets = [
-            DataSetType.CURRENT_WEATHER,
-            DataSetType.DAILY_FORECAST,
-            DataSetType.HOURLY_FORECAST,
-        ]
-
         self.supported_data_sets = [
             data_set
-            for data_set in requested_data_sets
+            for data_set in REQUESTED_DATA_SETS
             if data_set in supported_data_sets
         ]
-
         LOGGER.debug("Supported data sets: %s", self.supported_data_sets)
 
     async def _async_update_data(self):
```

bdraco reviewed (commented):
Some tiny review comments, and this should be good to merge.

tjhorner committed:
Improvements from code review (weatherkit)

tjhorner commented:
@bdraco Addressed your comments 👍

bdraco commented:
Sorry I missed it before, but one more review comment above.

bdraco commented:
Will also check with joostlek to see if he can do another pass in case I missed anything else

joostlek commented on homeassistant/components/weatherkit/weather.py:
```diff
@@ -0,0 +1,249 @@
+    @property
+    def current_weather(self) -> dict[str, Any]:
+        """Return current weather data."""
+        return self.coordinator.data.get("currentWeather")
```
```suggestion
        return self.data.get("currentWeather")
```

tjhorner commented on tests/components/weatherkit/test_coordinator.py:
```diff
@@ -0,0 +1,23 @@
+async def test_failed_updates(hass: HomeAssistant) -> None:
+    """Test that we properly handle failed updates."""
+    entry = await init_integration(hass)
+    coordinator = hass.data[DOMAIN][entry.entry_id]
```
How would I get a reference to the coordinator for this test? Or is there a better way to do this without a reference to it?

joostlek replied:
We don't want to test internals, we want to test what the result of something is. In other words, we want to test what happens when an update failed. To update, you can patch in a failed update and skip time to the next refresh. If the test is doing its job, you should see unavailable entities or unavailable and you compare that with what you expect.

joostlek replied:
I think if you search in the codebase for `async_fire_time_changed` you can find some examples

tjhorner replied:
Got it. I was thinking of advancing time artificially, but then I couldn't test that the exception was raised. But checking for unavailable entities makes sense.

tjhorner committed:
Modify coordinator test (weatherkit)

bdraco reviewed (approved):
Thanks @tjhorner

bdraco committed:
Merge branch 'dev' into apple-weatherkit

MartinHjelmare commented on homeassistant/components/weatherkit/config_flow.py:
```diff
@@ -0,0 +1,126 @@
+            location[CONF_LONGITUDE],
+        )
+
+        if len(availability) == 0:
```
```py
if not availability:
```

MartinHjelmare commented on homeassistant/components/weatherkit/const.py:
```diff
@@ -0,0 +1,13 @@
+
+NAME = "Apple WeatherKit"
+DOMAIN = "weatherkit"
+ATTRIBUTION = "Data provided by Apple Weather. https://developer.apple.com/weatherkit/data-source-attribution/"
```
Please break long strings around max 88 characters per line. I'd break here before the URL.

MartinHjelmare commented on homeassistant/components/weatherkit/weather.py:
```diff
@@ -0,0 +1,249 @@
+
+
+condition_code_to_hass = {
+    "BlowingDust": "windy",
```
Please import and use our weather conditions constants for the Home Assistant values:
https://github.com/home-assistant/core/blob/aff49cb67abc376d18a58347f13542bfbad81b6c/homeassistant/components/weather/__init__.py#L91-L105

MartinHjelmare commented on homeassistant/components/weatherkit/weather.py:
```diff
@@ -0,0 +1,249 @@
+}
+
+
+def _map_daily_forecast(forecast) -> Forecast:
```
Please type the whole signature when adding type annotations.

MartinHjelmare commented on homeassistant/components/weatherkit/weather.py:
```diff
@@ -0,0 +1,249 @@
+        if not self.coordinator.supported_data_sets:
+            return WeatherEntityFeature(0)
+
+        features = WeatherEntityFeature(0)
```
We could move this to the top and use on line 146.

MartinHjelmare commented on tests/components/weatherkit/test_config_flow.py:
```diff
@@ -0,0 +1,134 @@
+    assert result["errors"] == {}
+
+    with patch(
+        "homeassistant.components.weatherkit.config_flow.WeatherKitFlowHandler._test_config",
```
Don't patch our code in the config flow. Patch the client library.

MartinHjelmare commented on tests/components/weatherkit/test_config_flow.py:
```diff
@@ -0,0 +1,134 @@
+}
+
+
+async def _test_exception_generates_error(
```
This function is only used in one place. We can inline it.

MartinHjelmare commented on tests/components/weatherkit/test_setup.py:
```diff
@@ -0,0 +1,63 @@
+    ):
+        entry.add_to_hass(hass)
+        config_entries.current_entry.set(entry)
+        await async_setup_entry(hass, entry)
```
Don't use the integration function directly. Use `hass.config_entries.async_setup` and check the state of the config entry.

https://developers.home-assistant.io/docs/development_testing#writing-tests-for-integrations

MartinHjelmare reviewed (commented):
Please address the comments in a new PR. Thanks!
